Mathlib/InformationTheory/KullbackLeibler/Basic.lean

```diff
@@ -0,0 +1,238 @@
+/-
+Copyright (c) 2025 Rémy Degenne. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Rémy Degenne, Lorenzo Luccioli
+-/
+import Mathlib.InformationTheory.KullbackLeibler.KLFun
+import Mathlib.MeasureTheory.Decomposition.IntegralRNDeriv
+
+/-!
+# Kullback-Leibler divergence
+
+The Kullback-Leibler divergence is a measure of the difference between two measures.
+
+## Main definitions
+
+* `kl μ ν`: Kullback-Leibler divergence between two measures, with value in `ℝ≥0∞`,
+  defined as `∞` if `μ` is not absolutely continuous with respect to `ν` or
+  if the log-likelihood ratio `llr μ ν` is not integrable with respect to `μ`, and by
+  `ENNReal.ofReal (∫ x, llr μ ν x ∂μ + (ν univ).toReal - (μ univ).toReal)` otherwise.
+
+Note that our Kullback-Leibler divergence is nonnegative by definition (it takes value in `ℝ≥0∞`).
+However `∫ x, llr μ ν x ∂μ + (ν univ).toReal - (μ univ).toReal` is nonnegative for all finite
+measures `μ ≪ ν`, as proved in the lemma `integral_llr_add_sub_measure_univ_nonneg`.
+That lemma is our version of Gibbs' inequality ("the Kullback-Leibler divergence is nonnegative").
+
+## Main statements
+
+* `kl_eq_zero_iff` : the Kullback-Leibler divergence between two finite measures is zero if and only
+  if the two measures are equal.
+
+## Implementation details
+
+The Kullback-Leibler divergence on probability measures is `∫ x, llr μ ν x ∂μ` if `μ ≪ ν`
+(and the log-likelihood ratio is integrable) and `∞` otherwise.
+The definition we use extends this to finite measures by introducing a correction term
+`(ν univ).toReal - (μ univ).toReal`. The definition of the divergence thus uses the formula
+`∫ x, llr μ ν x ∂μ + (ν univ).toReal - (μ univ).toReal`, which is nonnegative for all finite
+measures `μ ≪ ν`. This also makes `kl μ ν` equal to an f-divergence: it equals the integral
+`∫ x, klFun (μ.rnDeriv ν x).toReal ∂ν`, in which `klFun x = x * log x + 1 - x`.
+
+-/
+
+open Real MeasureTheory Set
+
+open scoped ENNReal
+
+namespace InformationTheory
+
+variable {α : Type*} {mα : MeasurableSpace α} {μ ν : Measure α}
+
+open Classical in
+/-- Kullback-Leibler divergence between two measures. -/
+noncomputable def kl (μ ν : Measure α) : ℝ≥0∞ :=
```

Comment:
Maybe
```suggestion
noncomputable irreducible_def kl (μ ν : Measure α) : ℝ≥0∞ :=
```
as the definition is a little bit complicated and one should never want the kernel to unfold it.